designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,284 @@
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+### Drawbacks of requiring schemas
+
+Plugin authors who left out schemas will need to take the time to add them.
+
+Users of old/unmaintained plugins/rules which are missing schemas will not be able to use them until they are updated.
+
+### Drawbacks of requiring object-style rules
+
+Plugin authors who still have function-style rules will need to convert them to object-style rules.
+
+Users of old/unmaintained plugins/rules which still have function-style rules will not be able to use them until they are updated.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+This RFC describes a set of breaking changes that can be released in the next major version of ESLint. Both ESLint plugin authors and ESLint users will be affected and will be unable to use offending rules until they are updated. This will mostly affect a small number of older ESLint plugins that haven't been updated nor maintained in years.
+
+I did an analysis of the ESLint plugin ecosystem by writing a tool called [eslint-ecosystem-analyzer](https://github.com/bmish/eslint-ecosystem-analyzer) to find out how much of the ecosystem will be affected by these changes.
+
+For the top 100 ESLint plugins, 1% of total rules have options but are missing a schema (or said another way, 5% of rules with options are missing a schema). The percentages are only slightly higher (worse) when the top 1,000 ESLint plugins are considered (the top 1,000 results include many repositories that haven't been maintained nor used in years).
+
+For the top 100 ESLint plugins, 94% of rules are object rules. The percentage is only slightly lower (worse) when the top 1,000 ESLint plugins are considered.
+
+These results indicate that the vast majority of the ecosystem is already ready for these changes and will not be affected.
+
+While the current numbers are already supportive of making these changes, we have a number of means available to further prepare plugins for these changes, which we could optionally employ:
+
+- Communicating about these upcoming changes in blog posts leading up to the major release
+- Adding deprecation notices inside ESLint in a minor version release, showing a warning when executing linting for an offending rule:
+  > DEPRECATION WARNING: This rule has options but is missing `meta.schema` and will stop working in ESLint v9.
+  > The maintainer of this rule needs to release an updated version with a schema added.
+  >
+  > DEPRECATION WARNING: This rule is using the deprecated function-style format and will stop working in ESLint v9.
+  > The maintainer of this rule needs to release an updated version using object-style format.
```

Comment:
Is this necessary, since we will have deprecation warnings when running tests?

Showing warnings when executing linting could be annoying to end users, as there's nothing they can do to fix the issue.

Reply:
I was hoping that showing the deprecation warning to end users would result in them pinging the maintainer / filing an issue on github / or even opening a PR themselves with the fix. This would mainly be necessary for stagnant eslint plugins where the maintainer might not otherwise notice the deprecation warning since they aren't regularly running tests.

Perhaps I can propose a different timeline however:

1. When this RFC is approved, add the deprecation warning when running tests, in the hopes of most affected plugins seeing this.
2. In the final minor version of ESLint 8, add the deprecation warning for end users when running linting to try to catch any last holdouts.
3. In ESLint 9, change the deprecation warnings into fatal assertions.

Reply:
I'd vote for not showing warnings to end users since we have other ways to notify maintainers: deprecation warnings when running tests, blog posts, and prerelease versions. If all of that fails to notify some plugins, the worst case is that they'll be notified by users when we release ESLint 9 final.

Reply:
Agreed with @mdjermanovic. The prerelease versions should give users/plugins time to discover any remaining problems. Updated RFC to only add a deprecation warning when running tests.